File: clients/ios/MSCRemoteiOS_Swift/NetworkSafety.swift
import Foundation

enum NetworkSafety {
    /// Allow HTTP only when the host looks local/private (LAN/VPN ranges).
    /// URL.host already strips brackets from IPv6 literals (e.g. [::1] → ::1).
    static func isLocalOrPrivateHost(_ host: String) -> Bool {
        let h = host.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        if h == "localhost" || h == "127.0.0.1" { return true }
        if h.hasSuffix(".local") { return true }

        // Tailscale MagicDNS hostnames
        if h.hasSuffix(".ts.net") { return true }

        // IPv6 checks (before IPv4: colons in the string exclude IPv4 parsing)
        if let ipv6 = IPv6Address(h) {
            return ipv6.isPrivateOrLocal
        }

        // IPv4 checks
        if let ipv4 = IPv4Address(h) {
            return ipv4.isPrivateOrLocal
        }

        return false
    }

    static func httpIsAllowed(for url: URL) -> Bool {
        guard (url.scheme ?? "").lowercased() == "http" else { return true } // https is fine
        guard let host = url.host else { return false }
        return isLocalOrPrivateHost(host)
    }
}

private struct IPv6Address {
    let raw: String

    /// Matches IPv6 literals (contain colon, no dot — rejects IPv4-mapped ::ffff:a.b.c.d forms).
    init?(_ s: String) {
        guard s.contains(":"), !s.contains(".") else { return nil }
        self.raw = s.lowercased()
    }

    var isPrivateOrLocal: Bool {
        if raw == "::1" { return true }           // loopback (RFC 4291)
        if raw.hasPrefix("fe80:") { return true } // link-local (fe80::/10, RFC 4291)
        if raw.hasPrefix("fd") { return true }    // ULA (fd00::/8, RFC 4193)
        return false
    }
}

private struct IPv4Address {
    let a: Int
    let b: Int
    let c: Int
    let d: Int

    init?(_ s: String) {
        let parts = s.split(separator: ".").map(String.init)
        guard parts.count == 4 else { return nil }
        let nums = parts.compactMap { Int($0) }
        guard nums.count == 4 else { return nil }
        self.a = nums[0]; self.b = nums[1]; self.c = nums[2]; self.d = nums[3]
        guard (0...255).contains(a), (0...255).contains(b), (0...255).contains(c), (0...255).contains(d) else { return nil }
    }

    var isPrivateOrLocal: Bool {
        // RFC1918 private ranges
        if a == 10 { return true }
        if a == 192 && b == 168 { return true }
        if a == 172 && (16...31).contains(b) { return true }

        // Loopback
        if a == 127 { return true }

        // CGNAT range (often used by VPNs like Tailscale): 100.64.0.0/10
        if a == 100 && (64...127).contains(b) { return true }

        return false
    }
}

File: crates/msc-domain/src/network_safety.rs
//! Local/private host classification, used to decide whether plain HTTP
//! (rather than HTTPS) is safe to allow for a given host.
//!
//! Ported from `NetworkSafety.swift` (iOS) — but pinned to the verbatim copy
//! `NetworkSafetyTests.swift` carries as its own oracle (see that file's
//! header comment: "this suite pins a VERBATIM PORT of the iOS NetworkSafety
//! rule set... as the oracle instead"), not the iOS source's `IPv4Address`
//! struct directly. The two diverge on Tailscale's CGNAT range
//! (100.64.0.0/10): the iOS struct folds it into `isPrivateOrLocal`, but the
//! test suite's copy deliberately does not, since a separate function
//! (`preferredPairingHost`) handles `100.*` as a first-choice pairing host
//! instead. The 13 extracted fixtures (`fixtures/network-safety/`, from
//! P0.14) encode the test suite's behavior, so that's what this ports.

/// Allow HTTP only when the host looks local/private (LAN/VPN ranges).
///
/// `URL.host` always strips brackets from IPv6 literals (e.g. `[::1]` →
/// `::1`), so callers are expected to pass bare IPv6 forms.
pub fn is_local_or_private_host(host: &str) -> bool {
    let h = host.trim().to_lowercase();
    if h.is_empty() {
        return false;
    }
    if h == "localhost" || h == "127.0.0.1" {
        return true;
    }
    if h.ends_with(".local") {
        return true;
    }
    if h.ends_with(".ts.net") {
        return true;
    }

    // IPv6: colons present and no dots (excludes IPv4-mapped ::ffff:a.b.c.d forms).
    if h.contains(':') && !h.contains('.') {
        if h == "::1" {
            return true; // loopback (RFC 4291)
        }
        if h.starts_with("fe80:") {
            return true; // link-local (fe80::/10, RFC 4291)
        }
        if h.starts_with("fd") {
            return true; // ULA (fd00::/8, RFC 4193)
        }
        return false;
    }

    // IPv4 — verbatim port of the test suite's prefix-based checks.
    if h.starts_with("10.") {
        return true;
    }
    if h.starts_with("192.168.") {
        return true;
    }
    if h.starts_with("172.") {
        let parts: Vec<&str> = h.split('.').collect();
        if parts.len() >= 2
            && let Ok(second) = parts[1].parse::<i64>()
            && (16..=31).contains(&second)
        {
            return true;
        }
    }
    if h.starts_with("169.254.") {
        return true;
    }
    false
}
